noun/events.c

```diff
@@ -746,51 +750,57 @@ _ce_image_fine(u3e_image* img_u,
 }
 #endif
 
-/* u3e_save(): save current changes.
+/*
+  u3e_save(): save current changes.
+
+  First check that we are not in dry-run mode and that `_ce_patch_compose`
+  returns non-NULL. In either case, do nothing, otherwise update the
+  snapshot:
+
+  - Sync the image file.
+  - Delete the patchfile and free it.
+  - Write all dirty pages to disk; clear protection and dirty bits.
+  - Sync the patch files to disk
+  - Verify the patch - because why not?
+  - Write the patch data into the image file. (This is idempotent).
+  - Delete and free the patches.
+
+  XX What does this mean? "This has to block the main thread.
+  All further processing can happen in a separate thread, though we
+  can't save again till this completes."
```

Comment:
Once we've written the dirty pages to disk (and reset their dirty bits and protection flags), we could handle the rest of the checkpoint creation/processing in a new thread, or somehow in parallel with ongoing operations. As long as we don't try to save a new checkpoint until the in-progress checkpoint is complete.